discovery_engine_core/core/src/engine.rs

```diff
@@ -72,85 +74,134 @@ where
+            .collect::<FuturesUnordered<_>>()
+            .fold(vec![], |mut errors, stack| async move {
+                if let Err(error) = stack.write().await.rank(&self.ranker) {
+                    errors.push(Error::StackOpFailed(error));
+                }
+
+                errors
+            })
+            .await;
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::Errors(errors))
+        }
+    }
+
     /// Process the feedback about the user spending some time on a document.
-    pub fn time_logged(&mut self, time_logged: &TimeSpent) -> Result<(), Error> {
-        self.ranker.time_logged(time_logged)?;
+    pub async fn time_spent(&mut self, time_spent: &TimeSpent) -> Result<(), Error> {
+        self.ranker.time_spent(time_spent)?;
 
-        rank_stacks(self.stacks.values_mut(), &self.ranker)
+        self.rank_stacks().await
     }
 
     /// Process the feedback about the user reacting to a document.
-    pub fn user_reacted(&mut self, reacted: &UserReacted) -> Result<(), Error> {
-        let stack = self
-            .stacks
-            .get_mut(&reacted.stack_id)
-            .ok_or(Error::InvalidStackId(reacted.stack_id))?;
-
-        stack.update_relevance(reacted.reaction);
+    pub async fn user_reacted(&mut self, reacted: &UserReacted) -> Result<(), Error> {
+        self.stacks
+            .get(&reacted.stack_id)
+            .ok_or(Error::InvalidStackId(reacted.stack_id))?
+            .write()
+            .await
+            .update_relevance(reacted.reaction);
 
         self.ranker.user_reacted(reacted)?;
 
-        rank_stacks(self.stacks.values_mut(), &self.ranker)
+        self.rank_stacks().await
     }
-}
-
-/// The ranker could rank the documents in a different order so we update the stacks with it.
-fn rank_stacks<'a, R: Ranker>(
-    stacks: impl Iterator<Item = &'a mut Stack>,
-    ranker: &R,
-) -> Result<(), Error> {
-    let errors = stacks.fold(vec![], |mut errors, stack| {
-        if let Err(e) = stack.rank(ranker).map_err(Error::StackOpFailed) {
-            errors.push(e);
-        }
-
-        errors
-    });
 
-    if errors.is_empty() {
-        Ok(())
-    } else {
-        Err(Error::Errors(errors))
+    /// Updates the stacks with data related to the top key phrases of the current data.
+    #[allow(dead_code)]
+    async fn update_stacks(&mut self, top: usize) -> Result<(), Error> {
+        let key_phrases = &self.ranker.select_top_key_phrases(top);
+        let ranker = &self.ranker;
+
+        self.stacks
+            .values()
+            .map(|stack| async move {
+                stack
+                    .read()
+                    .await
+                    .ops
+                    .new_items(key_phrases, ranker)
+                    .await
+                    .map(|documents| (stack, documents))
+            })
+            .collect::<FuturesUnordered<_>>()
+            .try_for_each(|(stack, documents)| async move {
```

Comment:
Here we want to do as in `rank_stacks` and collect all the errors and return them.

Reply:
do we want to collect the errors from `new_items()` or from `update()` or from both?

Reply:
let's get everything. at the moment is our only way to tell outside that something happened